### PR TITLE
feat: add support for ipfs hashing

### PR DESCRIPTION
**Pull request by nchamo**
We are now adding support for v4 entities. This new version of entities will have IPFS v2 hashes instead of the current ones.

As part of this change we are also:
* Making snapshots use IPFS hashes
* Supporting wearable reverts with IPFS hashes

### CODE REVIEW

File: content/test/integration/E2ETestUtils.ts [hunk 1]
```diff
@@ -118,7 +118,7 @@ export async function deployEntitiesCombo(
   for (const { deployData } of entitiesCombo) {
     deploymentResult = await service.deployEntity(Array.from(deployData.files.values()), deployData.entityId, {
       authChain: deployData.authChain,
-      version: EntityVersion.V2
+      version: EntityVersion.V3
```
Comment (nchamo) [inline content/test/integration/E2ETestUtils.ts:121]: Should have been V3 all along

File: content/test/integration/service/deployments/deployment-filters.spec.ts [hunk 2]
```diff
@@ -122,7 +122,7 @@ describe('Integration - Deployment Filters', () => {
   async function deployWithAuditInfo(entities: EntityCombo[], overrideAuditInfo?: Partial<AuditInfo>) {
     const result: Timestamp[] = []
     for (const { deployData } of entities) {
-      const newAuditInfo = { version: EntityVersion.V2, authChain: deployData.authChain, ...overrideAuditInfo }
+      const newAuditInfo = { version: EntityVersion.V3, authChain: deployData.authChain, ...overrideAuditInfo }
```
Comment (nchamo) [inline content/test/integration/service/deployments/deployment-filters.spec.ts:125]: Should have been V3 all along

**Review by menduz — APPROVED** [review]
Awesome, thanks for adding the TODO comments to make next steps clearer

File: content/src/service/access/AccessCheckerForWearables.ts [hunk 3]
```diff
@@ -128,9 +128,9 @@ export class AccessCheckerForWearables {
           const entries = Array.from(content?.entries() ?? [])
           const contentAsJson = entries.map(([key, hash]) => ({ key, hash })).sort(compare)
           const buffer = Buffer.from(JSON.stringify({ content: contentAsJson, metadata }))
-          return Hashing.calculateBufferHash(buffer)
+          return Promise.all([Hashing.calculateBufferHash(buffer), Hashing.calculateIPFSHash(buffer)])
```
Comment (agusaldasoro) [inline content/src/service/access/AccessCheckerForWearables.ts:131]: Don't you have the entity version here so you can only execute the necessary algorithm?

File: content/src/service/snapshots/SnapshotManager.ts [hunk 4]
```diff
@@ -78,7 +78,7 @@ export class SnapshotManager {
           const buffer = Buffer.from(JSON.stringify(inArrayFormat))
 
           // Calculate the snapshot's hash
-          const hash = await Hashing.calculateBufferHash(buffer)
+          const hash = await Hashing.calculateIPFSHash(buffer)
```
Comment (agusaldasoro) [inline content/src/service/snapshots/SnapshotManager.ts:81]: Why are you sure that all the entities are v4?

File: content/test/unit/service/validations/Validations.spec.ts [hunk 5]
```diff
@@ -348,16 +348,56 @@ describe('Validations', function () {
       await assertNoErrors(result)
     })
   })
+
+  describe('IFPS hashing', () => {
+    it(`when an entity's id is not an ipfs hash, then it fails`, async () => {
```
Comment (agusaldasoro) [inline content/test/unit/service/validations/Validations.spec.ts:353]: Should we test also the older hashing algorithm so it fails?

File: content/src/service/access/AccessCheckerForWearables.ts (hunk 3 above)
Comment (nchamo) [reply content/src/service/access/AccessCheckerForWearables.ts:131]: But this hash is stored on the blockchain by the committee. I'm not sure which version they'll use, and it won't depend on the entity version. That's why we are checking for both

File: content/src/service/snapshots/SnapshotManager.ts (hunk 4 above)
Comment (nchamo) [reply content/src/service/snapshots/SnapshotManager.ts:81]: This has nothing to do with the entities themselves. The snapshot file will be exposed with the new hash, that's it. It doesn't matter if the entities are v3 or v4

File: content/test/unit/service/validations/Validations.spec.ts (hunk 5 above)
Comment (nchamo) [reply content/test/unit/service/validations/Validations.spec.ts:353]: Done!

**Review by agusaldasoro — APPROVED** [review]

**Pull request merged**